### ISSUE TITLE
oAuth Message Parser expects an int() where a string timestamp is present from Auth0

### ISSUE THREAD

**Issue by andrewkrug** [OpenIDC/pyoidc#417]
https://github.com/OpenIDC/pyoidc/blob/master/src/oic/oauth2/message.py#L450

Below is a regular auth0 profile.  Using the latest release of pyoidc v0.11.0.1 this profile fails to parse with the exception:

ValueError: "2017-08-28T16:10:58.969Z", wrong type of value for "updated_at"

I am assuming that is because the current validation expects an epoch time for update_at vs zulu time.

Please advise on the best course of action.

``` {
    "family_name": "Krug",
    "given_name": "Andrew",
    "nickname": "Andrew Krug",
    "groups": [
        "IntranetWiki",
        "StatsDashboard",
        "phonebook_access",
        "corp-vpn",
        "vpn_corp",
    ],
    "emails": [
        "akrug@mozilla.com"
    ],
    "dn": "mail=akrug@mozilla.com,o=com,dc=mozilla",
    "organizationUnits": "mail=akrug@mozilla.com,o=com,dc=mozilla",
    "email_verified": true,
    "email": "akrug@mozilla.com",
    "updated_at": "2017-08-28T16:10:58.969Z",
    "name": "Andrew Krug",
    "picture": "https://s.gravatar.com/avatar/abf4796da7aa08d55d5facd12be2c2b0?s=480&r=pg&d=https%3A%2F%2Fcdn.auth0.com%2Favatars%2Fak.png",
    "user_id": "ad|Mozilla-LDAP|akrug",
    "identities": [
        {
            "user_id": "Mozilla-LDAP|akrug",
            "provider": "ad",
            "connection": "Mozilla-LDAP",
            "isSocial": false
        }
    ],
    "created_at": "2017-02-06T22:53:30.473Z",
    "multifactor": [
        "duo"
    ],
    "_HRData": {
        "cost_center": "1420 - Enterprise Information Security (EIS)",
        "management_level": "Individual Contributor",
        "manager_name": "Jeff Bryner",
        "manager_email": "jbryner@mozilla.com",
        "business_title": "Security Engineer: Identity and Access Management",
        "test_value": "empty"
    },
    "user_metadata": {},
    "app_metadata": {
        "lastName": "Krug",
        "userName": "akrug",
        "firstName": "Andrew",
        "displayName": "Andrew J Krug",
        "active": true,
        "shirtSize": "Fitted Small",
        "preferredLanguage": "en_US",
        "timezone": "US/Pacific",
        "primaryEmail": "akrug@mozilla.com",
        "lastModified": "2017-08-18T16:48:58+00:00",
        "created": "2017-02-06T22:53:54+00:00",
        "picture": "https://cdn.mozillians.org/media/uploads/sorl-cache/92/2e/922efe09c26b1a04c93b7580ffae866d.jpg",
        "uris": [
            {
                "verified": false,
                "name": "mozillians-GitHub-39660",
                "value": "https://github.com/andrewkrug",
                "primary": false
            }
        ],
        "groups": [
            "mozilliansorg_iam-project",
            "mozilliansorg_cloud incident response",
            "mozilliansorg_cis_whitelist",
            "mozilliansorg_agile",

        ],
        "emails": [
            {
                "verified": true,
                "name": "mozillians-primary",
                "value": "akrug@mozilla.com",
                "primary": true
            }
        ]
    },
    "last_ip": "208.186.8.92",
    "last_login": "2017-08-28T16:10:58.969Z",
    "logins_count": 429,
    "blocked_for": [],
    "guardian_enrollments": []
}
```

**Comment by andrewkrug** [comment]
OIDC standard says: 

updated_at | number | Time the End-User's information was last updated. 	    Its value is a JSON number representing the number of seconds from 	    1970-01-01T0:0:0Z as measured in UTC until the date/time.

So in theory this case should work.

**Comment by schlenk** [comment]
I would say Auth0 is violating the spec as you cited it. They put a string inside but it should be seconds since the Unix epoche.

They probably thought it is easier that way, as all the private claims they added are in the same format.

**Comment by gdestuynder** [comment]
Unfortunately yes - as per http://openid.net/specs/openid-connect-implicit-1_0.html#StandardClaims for example `updated_at` should be number.
Even in OIDC Conformant mode, Auth0 uses a string.

**Comment by rohe** [comment]
Sounds clear to me, it MUST be a number.

**Comment by lwm** [comment]
So, if I read this right, this is on the side of Auth0, not us. 

Closing off, please re-open if I got it wrong.

**Comment by jgehrcke** [comment]
For the record: Auth0 sill enhance compatibility for you, if you ask them. See Auth0's response to my inquiry here: https://community.auth0.com/t/oidc-id-token-claim-updated-at-violates-oidc-specification-breaks-rp-implementations/24098

**Comment by gdestuynder** [comment]
Another work-around is a rule that fixes it in Auth0, we used that for a while, if that helps anyone: https://github.com/mozilla-iam/auth0-deploy/commit/0aa84fd413eee36843d421bc28918093daf0998a#diff-a4f86f7bfc24194b276c22e0ef158197

tldr:
```
context.idToken.updated_at = Math.floor(new Date(user.updated_at)/1000);
```

**Comment by lafkuku** [comment]
I also ran into this issue with OneLogin. We reported the issue and they are fixing similar to Auth0. However, I ended up creating a slightly different work-around until it's fixed by hooking the deserialization flow. 

`
def updated_at_deserializer(val, sformat="dict"):
    utc_time = datetime.datetime.strptime(val, "%Y-%m-%dT%H:%M:%SZ")
    epoch_time = (utc_time - datetime.datetime(1970, 1, 1)).total_seconds()
    return int(epoch_time)

UPDATE_AT_PARAM = ParamDefinition(int, False, None, updated_at_deserializer, False)
IdToken.c_param["updated_at"] = UPDATE_AT_PARAM
`

**Comment by alehuo** [comment]
Issue still exists as of 20.12.2019 with pyoidc 1.1.2

**Comment by tpazderka** [comment]
This is not an issue with pyoidc, but with your provider. The specs state that the timestamp MUST be a number, so sending anything else is a violation of specs.

**Comment by hmoffatt** [comment]
This still isn't fixed at Auth0. I tried @lafkuku's suggestion but it doesn't quite fit Auth0; their timestamp includes milliseconds as well (which strptime() can't parse directly). Also the user info also includes the faulty updated_at field.

Since I do not need the updated_at field I deleted it from OpenIDSchema and IdToken instead.
